File: leetcode/problems/problem-specific/128.longest-consecutive-sequence/128.longest-consecutive-sequence.go
package main

/*
 * @lc app=leetcode id=128 lang=golang
 *
 * [128] Longest Consecutive Sequence
 */

// @lc code=start

// * HashSet and Intelligent Sequence Building Solution | O(n) Time | O(n) Space
// * n -> Length of the nums array

func longestConsecutive(nums []int) int {
	numsSize := len(nums)
	if numsSize <= 1 {
		return numsSize
	}

	var exists = struct{}{}
	mappings, longestConsecutiveSequence := make(map[int]struct{}), 0
	for _, num := range nums {
		mappings[num] = exists
	}

	for _, num := range nums {
		// * Check if it's the start of the sequence
		if _, ok := mappings[num-1]; !ok {
			curElement, curSequenceSize := num, 1
			for {
				if _, ok := mappings[curElement+1]; ok {
					curElement += 1
					curSequenceSize += 1
				} else {
					break
				}
			}

			longestConsecutiveSequence = findMax(longestConsecutiveSequence, curSequenceSize)
		}
	}

	return longestConsecutiveSequence
}

func findMax(a, b int) int {
	if a >= b {
		return a
	}

	return b
}

// @lc code=end

File: leetcode/problems/sessions/2025-10-25_2025-??-??/128.longest-consecutive-sequence.py
#
# @lc app=leetcode id=128 lang=python3
#
# [128] Longest Consecutive Sequence
#

# @lc code=start


# * Hash Table and Intelligent Sequence Building Solution | O(n) Time | O(n) Space
# * n -> The length of nums array


class Solution:
    def longestConsecutive(self, nums: List[int]) -> int:
        nums_set = set(nums)
        longest_sequence = 0
        for num in nums_set:
            if (num - 1) not in nums_set:
                cur_sequence = 1
                while (num + cur_sequence) in nums_set:
                    cur_sequence += 1
                longest_sequence = max(longest_sequence, cur_sequence)

        return longest_sequence


# @lc code=end
